File: my-folder/0169-majority-element/solution.c
int majorityElement(int* nums, int numsSize) {
    // The Boyer-Moore Voting Algorithm
    int candidate = nums[0];
    int count = 1;

    // Phase 1: Find a candidate for the majority element
    for (size_t i = 1; i < numsSize; i++) {
        if (nums[i] == candidate) {
            count++;
        } else {
            count--;
            if (count == 0) {
            candidate = nums[i];
            count = 1;
            }
        }
    }

    // Phase 2: Verify that the candidate is indeed the majority element
    count = 0;
    for (size_t i = 0; i < numsSize; i++) {
        if (nums[i] == candidate) {
            count++;
        }
    }

    // Since the problem guarantees that a majority element always exists,
    // we do not need to check if count > numsSize / 2
    return candidate;
}

File: my-folder/0169-majority-element/solution.rs
impl Solution {
    pub fn majority_element(nums: Vec<i32>) -> i32 {
        // The Boyer-Moore Voting Algorithm
        // Phase 1: Find a candidate for the majority element
        let mut candidate = nums[0];
        let mut count = 1;

        nums.iter().skip(1).for_each(|&num| {
            if num == candidate {
                count += 1;
            } else {
                count -= 1;
                if count == 0 {
                    candidate = num;
                    count = 1;
                }
            }
        });

        // Phase 2: Verify that the candidate is indeed the majority element
        count = nums.iter().filter(|&&num| num == candidate).count() as i32;

        // Assuming the majority element always exists
        candidate
    }
}
